cpp/docs/TRANSITIONGUIDE.md

```diff
@@ -0,0 +1,206 @@
+
+## Namespaces
+
+All external APIs should be placed in the `cudf` namespace. For example:
+
+```
+namespace cudf{
+
+  void some_function(...);
+
+} // namespace cudf
+
+```
+
+### `detail` Namespace
+
+Utility functions that are for use only in libcudf internals, but can be reused in many places should be placed in the `cudf::detail` namespace. For example:
+
+```
+namespace cudf{
+namespace detail{
+void reusable_utility_function(...);
+} // namespace detail
+} // namespace cudf
+```
+
+### Anonymous Namespaces
+
+All functions specific to a translation unit should be placed in an anonymous namespace.
+
+Example:
+```
+namespace {
+void some_specific_helper_function(...);
+} // namespace
+```
+
+## Data Structures
+
+### `cudf::column`
+
+A replacement abstraction for `gdf_column` is actively being designed: https://github.com/rapidsai/cudf/issues/1443
+
+Until that design is complete, continue to use `gdf_column`s as usual. 
+
+### `cudf::table`
+
+It is common for an API to work on a set of `N` `gdf_column`s of equal size which required an API to do something like the following: 
+
+```
+void some_multi_col_function(gdf_column * cols[], int num_cols);
+```
+
+A simple wrapper class called `cudf::table` is defined in `cudf/cpp/src/table/table.hpp` to provide a convenient abstraction for the above use case. This class should be used any time a function is operating on a collection of `gdf_column`s of equal size.
+
+For example, the above becomes:
+```
+void some_multi_col_function(cudf::table & table);
+```
+
+## Input/Output Style
+
+The preferred style for how inputs are passed in and outputs are returned is the following:
+
+* Inputs 
+   * Expensive to copy types --- pass by constant reference
+   * Inexpensive to copy types --- pass by value
+* In/Outs - Pass via pointer
+* Outputs - Returned by value
+   * Returning outputs by value will not incur a copy in most cases. For more details, see: https://stackoverflow.com/questions/12953127/what-are-copy-elision-and-return-value-optimization
+   * **NOTE:** This may mean returning a `gdf_column` whose device memory was allocated within a `libcudf` function and it is the caller's responsibility to free the underlying memory. This is a temporary headache until a `cudf::column` abstraction is designed to allieviate it. 
+
+For example:
+
+```
+namespace cudf{
+column some_function( cudf::table const& input_table, cudf::table * input_output_table, int size){
+  column col;
+  // Do stuff to create output col
+  ...
+  // Return the result column of this function
+  return col;
+}
+}
+```
+
+And the usage would be:
+```
+cudf::table input;
+cudf::table in_out;
+int size;
+column output = cudf::some_function(input, &in_out, size);
+```
+
+### Multiple Return Values
+
+Sometimes it is necessary for functions to have multiple outputs. There are a few ways this can be done in C++ (including creating a `struct` for the output). One convenient way to do this is using `std::tie` and `std::make_tuple`.
```

Comment:
`std::tie` is on the side of the code _getting_ multiple values - and it's not always necessary, it's just one option. You could very well write
```
auto foo = cudf::return_multiple_outputs();
auto& bar = std::get<0>(foo);
auto& baz = std::get<1>(foo);
```